File: 2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        vector<int> v;
        int count=1;
        while(head->next->next!=NULL){
            if(head->val>head->next->val && head->next->val<head->next->next->val || head->val<head->next->val && head->next->val>head->next->next->val){
                v.push_back(count+1);
            }
            head=head->next;
            count++;
        }
        for(auto i:v){
            cout<<i<<" ";
        }
        if(v.size()==0) return{-1,-1};
        sort(v.begin(),v.end());
        int m=v[v.size()-1]-v[0];
        int mi=INT_MAX;
        for(int i=0;i<v.size()-1;i++){
            mi=min(mi,v[i+1]-v[i]);
        }
        if(m==0 || mi==INT_MAX) return {-1,-1};
        return{mi,m};
    
    }
};

File: 2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public int[] NodesBetweenCriticalPoints(ListNode head) {
        ListNode cur=head.next;
        ListNode prev=head;
        List<int> li=new List<int>();
        int count=1;
        while(cur.next!=null){
            ListNode nxt=cur.next;
            if(prev.val<cur.val && cur.val>nxt.val) li.Add(count);
            if(prev.val>cur.val && nxt.val>cur.val) li.Add(count);
            count++;
            cur=cur.next;
            prev=prev.next;
        }
        if(li.Count<2) return new int[]{-1,-1};
        li.Sort();
        int minDis=int.MaxValue;
        for(int i=0;i<li.Count-1;i++){
            minDis=Math.Min(li[i+1]-li[i],minDis);
        }
        int maxDis=li[li.Count-1]-li[0];
        return new int[]{minDis,maxDis};

    }
}